From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Markus Rudy <mr@edgeless.systems>
Date: Tue, 9 Jun 2026 12:57:42 +0200
Subject: [PATCH] runtime: don't attempt to clean up cgroup scope

The cleanup (a systemctl stop) is asynchronous by nature, but the
containerd cgroup code always waits for its completion without
considering context cancellation etc. This can lead to errors not being
visible at the Kubernetes level if the scope unit is not stopped before
the shim is terminated. The result is just a "create container timeout"
message.

It turns out that removing the scope is optional: containerd will remove
the parent scope when cleaning up, which results in our scope being
cleaned up, too.

Included in this commit are two error propagation fixes which should
make problems with cgroups apparent much earlier. The cgroupv1 code path
has a similar error, but is not checked because we can't test it.
Furthermore, the error message mentioned above is improved such that
it's clearer why we're returning.

Signed-off-by: Markus Rudy <mr@edgeless.systems>
---
 src/runtime/pkg/containerd-shim-v2/service.go |  2 +-
 src/runtime/pkg/resourcecontrol/cgroups.go    | 11 ++++++-----
 src/runtime/virtcontainers/sandbox.go         |  4 ++++
 3 files changed, 11 insertions(+), 6 deletions(-)

diff --git a/src/runtime/pkg/containerd-shim-v2/service.go b/src/runtime/pkg/containerd-shim-v2/service.go
index c9fad70d0540a3667cc237e2af97ef90c03d90b1..7f4f430782f05a8e9c7775f27c9d7adacda29e4f 100644
--- a/src/runtime/pkg/containerd-shim-v2/service.go
+++ b/src/runtime/pkg/containerd-shim-v2/service.go
@@ -446,7 +446,7 @@ func (s *service) Create(ctx context.Context, r *taskAPI.CreateTaskRequest) (_ *
 
 	select {
 	case <-ctx.Done():
-		return nil, errors.Errorf("create container timeout: %v", r.ID)
+		return nil, errors.Errorf("create container RPC expired: %v (id: %s)", ctx.Err(), r.ID)
 	case res := <-ch:
 		if res.err != nil {
 			return nil, res.err
diff --git a/src/runtime/pkg/resourcecontrol/cgroups.go b/src/runtime/pkg/resourcecontrol/cgroups.go
index d5a17306282c7375103e030ab23e2869db37698a..d4510a66cbe79ba6558a794d5c74d9676d32f327 100644
--- a/src/runtime/pkg/resourcecontrol/cgroups.go
+++ b/src/runtime/pkg/resourcecontrol/cgroups.go
@@ -214,8 +214,8 @@ func NewSandboxResourceController(path string, resources *specs.LinuxResources,
 	//v1 and v2 cgroups against systemd, the following interacts directly with systemd
 	//to create the cgroup and then load it using containerd's api.
 	//adding runtime process, it makes calling setupCgroups redundant
-	if createCgroupsSystemd(slice, unit, os.Getpid()); err != nil {
-		return nil, err
+	if err := createCgroupsSystemd(slice, unit, os.Getpid()); err != nil {
+		return nil, fmt.Errorf("creating systemd cgroup: %w", err)
 	}
 
 	// Create systemd cgroup
@@ -237,9 +237,10 @@ func NewSandboxResourceController(path string, resources *specs.LinuxResources,
 		// load created cgroup and update with resources
 		cg, err := cgroup2.LoadSystemd(slice, unit)
 		if err != nil {
-			if cg.Update(cgroup2.ToResources(&sandboxResources)); err != nil {
-				return nil, err
-			}
+			return nil, fmt.Errorf("loading systemd cgroupv2: %w", err)
+		}
+		if err := cg.Update(cgroup2.ToResources(&sandboxResources)); err != nil {
+			return nil, fmt.Errorf("updating systemd cgroupv2: %w", err)
 		}
 		cgroup = cg
 	} else {
diff --git a/src/runtime/virtcontainers/sandbox.go b/src/runtime/virtcontainers/sandbox.go
index c96885dadc7d0d8c3555db0ab0d24ecbb35eba4a..111c3da3de280c421349a6444352e5eb32174775 100644
--- a/src/runtime/virtcontainers/sandbox.go
+++ b/src/runtime/virtcontainers/sandbox.go
@@ -2874,6 +2874,10 @@ func (s *Sandbox) resourceControllerUpdate(ctx context.Context) error {
 // resourceControllerDelete will move the running processes in the sandbox resource
 // cvontroller to the parent and then delete the sandbox controller.
 func (s *Sandbox) resourceControllerDelete() error {
+	if resCtrl.IsSystemdCgroup(s.state.SandboxCgroupPath) {
+		s.Logger().Debug("Skipping resource controller deletion on systemd platform")
+		return nil
+	}
 	s.Logger().Debugf("Deleting sandbox %s resource controler", s.sandboxController)
 	if s.state.SandboxCgroupPath == "" {
 		s.Logger().Warnf("sandbox %s resource controler path is empty", s.sandboxController)
